### PR TITLE
Invert the default output parsing for TextGenerator subtypes

### PR DESCRIPTION
**Pull request by BenWilson2**
## Related Issues/PRs

 #xxx

## What changes are proposed in this pull request?

Inverts the output parsing default configuration making newline removal and prompt stripping opt-in via an inference_config setting.

## How is this patch tested?

- [X] Existing unit/integration tests
- [X] New unit/integration tests
- [ ] Manual tests (describe details, including test results, below)

## Does this PR change the documentation?

- [X] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly in the documentation preview.

### Is this a user-facing change?

- [X] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [X] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [X] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/transformers.py [hunk 1]
```diff
@@ -1647,8 +1650,8 @@ def trim_input(data_in, data_out):
                     data_out = data_out[len(data_in) :].lstrip()
                     if data_out.startswith("A:"):
                         data_out = data_out[2:].lstrip()
-                for to_replace, replace in replacements.items():
-                    data_out = data_out.replace(to_replace, replace)
+                    for to_replace, replace in replacements.items():
+                        data_out = data_out.replace(to_replace, replace)
```
Comment (harupy) [inline mlflow/transformers.py:1654]: What if a user doesn't want the prompt, but wants to preserve new characeters in the output?

Comment (BenWilson2) [reply mlflow/transformers.py:1654]: I can add an additional kwargs entry for that

File: mlflow/transformers.py [hunk 2]
```diff
@@ -1609,13 +1616,19 @@ def _coerce_exploded_dict_to_single_dict(self, data):
...
+        flavor_config,
+        include_prompt=True,
+        remove_newlines=False,
     ):
         """
         Parse the output from instruction pipelines to conform with other text generator
         pipeline types and remove line feed characters and other confusing outputs
         """
-        replacements = {"\n\n": " "}
+        replacements = {"\n\n": " ", "\n": " "}
```
Comment (harupy) [inline mlflow/transformers.py:1631]: I'd use a regular expression here to shrink multiple newline characeters into a space.

File: docs/source/models.rst [hunk 3]
```diff
@@ -2000,6 +2000,15 @@ and standardizes both the inputs and outputs of pipeline inference. This conform
 and batch inference by coercing the data structures that are required for ``transformers`` inference pipelines
 to formats that are compatible with json serialization and casting to Pandas DataFrames.
 
+.. note::
+    Certain `TextGenerationPipeline` types, particularly instructional-based ones, may return the original
+    prompt and included line-formatting carriage returns `"\n"` in their outputs. For these pipeline types,
+    if you would like to disable the prompt return, you can set the following in the `inference_config` dictionary when
+    saving or logging the model: `"include_prompt": False`. To remove the newline characters from within the body
+    of the generated text output, you can add the `"remove_newlines": True` option to the `inference_config` dictionary.
```
Comment (harupy) [inline docs/source/models.rst:2008]: How about the option name like `shrink_newlines`?

File: mlflow/transformers.py [hunk 4]
```diff
@@ -1446,9 +1446,16 @@ def _predict(self, data):
...
             )
 
-        # Optional input preservation for specific pipeline types
+        # Optional input preservation for specific pipeline types. This is True (include raw
+        # formatting output), but if `include_prompt` is set to False in the `inference_config`
+        # option during model saving, excess newline characters and the fed-in prompt will be
+        # trimmed out from the start of the response.
         include_prompt = (
-            self.inference_config.pop("include_prompt", False) if self.inference_config else False
+            self.inference_config.pop("include_prompt", True) if self.inference_config else True
```
Comment (harupy) [inline mlflow/transformers.py:1454]: Does `include_prompt` need to be popped out?

File: docs/source/models.rst (hunk 3 above)
Comment (harupy) [reply docs/source/models.rst:2008]: Other candidates (rejected in me):

- `replace_newlines`: makes me wonder "replace with what?"
- `replace_newlines_with_space`: clear but too long

File: mlflow/transformers.py (hunk 4 above)
Comment (BenWilson2) [reply mlflow/transformers.py:1454]: unfortunately, yes. 
transformers pipeline execution includes validation for kwargs submitted. If we leave that inference kwarg entry in (by using `self.inference_config.get(...)`, we get:

E           ValueError: The following `model_kwargs` are not used by the model: ['include_prompt'] (note: typos in the generate arguments will also show up in this list)

File: docs/source/models.rst (hunk 3 above)
Comment (BenWilson2) [reply docs/source/models.rst:2008]: I like it :) changing!

File: mlflow/transformers.py (hunk 2 above)
Comment (BenWilson2) [reply mlflow/transformers.py:1631]: great point. Updated and added a spacing collapse for a weird edge case that can happen

File: mlflow/transformers.py [hunk 5]
```diff
@@ -1609,13 +1617,19 @@ def _coerce_exploded_dict_to_single_dict(self, data):
...
+        flavor_config,
+        include_prompt=True,
+        shrink_newlines=False,
     ):
         """
         Parse the output from instruction pipelines to conform with other text generator
         pipeline types and remove line feed characters and other confusing outputs
         """
-        replacements = {"\n\n": " "}
+        replacements = {"\n+": " ", "\\s+": " "}
```
Comment (harupy) [inline mlflow/transformers.py:1632]: Is `\\s+` a newline?

Comment (harupy) [reply mlflow/transformers.py:1632]: btw why do we need to replace `\\s+`? Does Dolly produce a response like `I    am Dolly`?

Comment (BenWilson2) [reply mlflow/transformers.py:1632]: If we're not declaring the match condition as a raw string (i.e., `r"\s+"`, which looks odd as a dict key), then the literal escape sequence is equivalent if using either a single `\s+` or double `\\s+`. 

example:

```
data = "Just\n\n testing\n something\n\n out\n\n\nhere.\n\n"
print("raw:")
print(data)
data = re.sub("\n+", " ", data)
print("remove newlines:")
print(data)
data_single = re.sub("\s+", " ", data)
print("remove extra spaces:")
print(data_single)
data_double = re.sub("\\s+", " ", data)
print("remove extra spaces double escape:")
print(data_double)

assert data_single == data_double
```

outputs:

```
raw:
Just

 testing
 something

 out


here.


remove newlines:
Just  testing  something  out here. 
remove extra spaces:
Just testing something out here. 
remove extra spaces double escape:
Just testing something out here. 
```

**Review by dbczumar — APPROVED** [review]
LGTM! Thanks @BenWilson2 !

**Review by harupy — APPROVED** [review]
LGTM once we address what we discussed in the standup!

**Comment by BenWilson2** [comment]
Adding: convert to `collapse_whitespace`, additional NB notes that make it clear when these are being used (why and what their defaults are), docstring for the method, and updating the docs.

**Pull request merged**